core/embed/trezorhal/stm32u5/haptic/drv2625/drv2625.c

```diff
@@ -0,0 +1,176 @@
+#endif
+
+void set_reg(uint8_t addr, uint8_t value) {
+  uint8_t data[] = {addr, value};
+  i2c_transmit(DRV2625_I2C_INSTANCE, I2C_ADDRESS, data, sizeof(data), 1);
+}
+
+void read_reg(uint8_t addr, uint8_t *value) {
+  i2c_transmit(DRV2625_I2C_INSTANCE, I2C_ADDRESS, &addr, 1, 1);
+  i2c_receive(DRV2625_I2C_INSTANCE, I2C_ADDRESS, value, 1, 1);
+}
+
+void haptic_calibrate(void) {
+  set_reg(DRV2625_REG_MODE, DRV2625_REG_MODE_AUTOCAL);
+  HAL_Delay(1);
+  set_reg(DRV2625_REG_GO, DRV2625_REG_GO_GO);
+
+  HAL_Delay(3000);
+}
+
+void haptic_init(void) {
+  // select library
+  set_reg(DRV2625_REG_LIBRARY, LIB_SEL | DRV2625_REG_LIBRARY_GAIN_25);
+  set_reg(DRV2625_REG_LRAERM,
+          LRA_ERM_SEL | LOOP_SEL | DRV2625_REG_LRAERM_AUTO_BRK_OL);
+
+  set_reg(DRV2625_REG_OD_CLAMP, ACTUATOR_OD_CLAMP);
+
+  set_reg(DRV2625_REG_LRA_WAVE_SHAPE, DRV2625_REG_LRA_WAVE_SHAPE_SINE);
+
+  set_reg(DRV2625_REG_OL_LRA_PERIOD_LO, ACTUATOR_LRA_PERIOD & 0xFF);
+  set_reg(DRV2625_REG_OL_LRA_PERIOD_HI, ACTUATOR_LRA_PERIOD >> 8);
+
+  GPIO_InitTypeDef GPIO_InitStructure;
+  GPIO_InitStructure.Mode = GPIO_MODE_AF_PP;
+  GPIO_InitStructure.Pull = GPIO_PULLDOWN;
+  GPIO_InitStructure.Speed = GPIO_SPEED_FREQ_LOW;
+  GPIO_InitStructure.Pin = GPIO_PIN_8;
+  GPIO_InitStructure.Alternate = GPIO_AF14_TIM16;
+  HAL_GPIO_Init(GPIOB, &GPIO_InitStructure);
+
+  TIM_HandleTypeDef TIM_Handle;
+  __HAL_RCC_TIM16_CLK_ENABLE();
+  TIM_Handle.State = HAL_TIM_STATE_RESET;
+  TIM_Handle.Instance = TIM16;
+  TIM_Handle.Init.Period = 0;
+  TIM_Handle.Init.Prescaler = SystemCoreClock / 10000;
+  TIM_Handle.Init.ClockDivision = TIM_CLOCKDIVISION_DIV1;
+  TIM_Handle.Init.CounterMode = TIM_COUNTERMODE_UP;
+  TIM_Handle.Init.RepetitionCounter = 0;
+  HAL_TIM_PWM_Init(&TIM_Handle);
+
+  TIM_OC_InitTypeDef TIM_OC_InitStructure;
+  TIM_OC_InitStructure.OCMode = TIM_OCMODE_PWM2;
+  TIM_OC_InitStructure.OCPolarity = TIM_OCPOLARITY_HIGH;
+  TIM_OC_InitStructure.Pulse = 1;
+  TIM_OC_InitStructure.OCNPolarity = TIM_OCNPOLARITY_HIGH;
+  TIM_OC_InitStructure.OCFastMode = TIM_OCFAST_DISABLE;
+  HAL_TIM_PWM_ConfigChannel(&TIM_Handle, &TIM_OC_InitStructure, TIM_CHANNEL_1);
+
+  HAL_TIM_OC_Start(&TIM_Handle, TIM_CHANNEL_1);
+
+  TIM16->CR1 |= TIM_CR1_OPM;
+  TIM16->BDTR |= TIM_BDTR_MOE;
+}
+
+void haptic_play_RTP(uint8_t value, uint16_t duration_ms) {
+  set_reg(DRV2625_REG_MODE,
+          DRV2625_REG_MODE_RTP | DRV2625_REG_MODE_TRGFUNC_ENABLE);
+  set_reg(DRV2625_REG_RTP, value);
+
+  if (duration_ms > 6500) {
+    duration_ms = 6500;
+  }
+
+  TIM16->CNT = 1;
+  TIM16->CCR1 = 1;
+  TIM16->ARR = duration_ms * 10;
+  TIM16->CR1 |= TIM_CR1_CEN;
+}
+
+void haptic_play_lib(lib_effect_t effect) {
+  set_reg(DRV2625_REG_MODE, DRV2625_REG_MODE_WAVEFORM);
+  set_reg(DRV2625_REG_WAVESEQ1, effect);
+  set_reg(DRV2625_REG_WAVESEQ2, 0);
+  set_reg(DRV2625_REG_GO, DRV2625_REG_GO_GO);
+}
+
+void haptic_play(haptic_effect_t effect) {
+  switch (effect) {
+    case HAPTIC_BUTTON_PRESS:
+      haptic_play_RTP(25, 10);
```

Comment:
added the defines (also for duration): https://github.com/trezor/trezor-firmware/commit/81406b9167c7e244310fd53c1f0cda9b84ac8e20

these are set experimentally to what feels good to me, though the effects are to be reviewed by designers. I would say that the specific setting is dependent on the actuator, board and mechanical properties, so it does not exactly belongs to the driver implementation. I can move it out to board.h, but for now as we have 1 device with haptic i thought it might stay here and we can deal with it later. Maybe we can refactor the boards mechanism in the meantime to be more sustainable, as discussed earlier today.